runtime/interpreter/interpreter.go

```diff
@@ -1614,181 +1616,144 @@ func (interpreter *Interpreter) VisitIdentifierExpression(expression *ast.Identi
 	return Done{Result: variable.Value}
 }
 
-// valueTuple
-
-type valueTuple struct {
-	left, right Value
-}
-
-// visitBinaryOperation interprets the left-hand side and the right-hand side and returns
-// the result in a valueTuple
-func (interpreter *Interpreter) visitBinaryOperation(expr *ast.BinaryExpression) Trampoline {
-	// interpret the left-hand side
-	return expr.Left.Accept(interpreter).(Trampoline).
-		FlatMap(func(left interface{}) Trampoline {
-			// after interpreting the left-hand side,
-			// interpret the right-hand side
-			return expr.Right.Accept(interpreter).(Trampoline).
-				FlatMap(func(right interface{}) Trampoline {
-					tuple := valueTuple{
-						left.(Value),
-						right.(Value),
-					}
-					return Done{Result: tuple}
-				})
-		})
-}
-
-func (interpreter *Interpreter) visitNumberBinaryOperation(
-	expression *ast.BinaryExpression,
-	f func(left, right NumberValue) Value,
-) ast.Repr {
-	return interpreter.visitBinaryOperation(expression).
-		Map(func(result interface{}) interface{} {
-			tuple := result.(valueTuple)
-			left := tuple.left.(NumberValue)
-			right := tuple.right.(NumberValue)
-			return f(left, right)
-		})
+func (interpreter *Interpreter) visitBinaryOperation(
+	expr *ast.BinaryExpression,
+	f func(left, right Value) Value,
+) Trampoline {
+	left := interpreter.runAllStatements(expr.Left.Accept(interpreter).(Trampoline)).(Value)
+	right := interpreter.runAllStatements(expr.Right.Accept(interpreter).(Trampoline)).(Value)
+	return Done{Result: f(left, right)}
 }
 
 func (interpreter *Interpreter) VisitBinaryExpression(expression *ast.BinaryExpression) ast.Repr {
 	switch expression.Operation {
 	case ast.OperationPlus:
-		return interpreter.visitNumberBinaryOperation(
+		return interpreter.visitBinaryOperation(
```

Comment:
I am thinking maybe we could do another minor optimization (also reduces code) here, by avoiding the creation of function pointers and avoiding the delegation.
```go
func (interpreter *Interpreter) VisitBinaryExpression(expression *ast.BinaryExpression) ast.Repr {
	left := interpreter.evaluate(expression.Left)
	right := interpreter.evaluate(expression.Right)

	var result Value

	switch expression.Operation {
	case ast.OperationPlus:
		result = left.(NumberValue).Plus(right.(NumberValue))
	case ast.OperationMinus:
		result = left.(NumberValue).Minus(right.(NumberValue))
	case ast.OperationMod:
		result = left.(NumberValue).Mod(right.(NumberValue))
	}

	return Done{Result: result}
}

```

we could early exit the ones that are not migrated yet.

Reply:
Yep, in the final PR the code looks similar to this 👍